Looks like this fails for some dtypes. From Travis:
```
======================================================================
ERROR: numpy.core.tests.test_numeric.TestNonzero.test_count_nonzero_axis_all_dtypes
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_numeric.py", line 1070, in test_count_nonzero_axis_all_dtypes
    assert_equal_w_dt(np.count_nonzero(m, axis=0),
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.6/site-packages/numpy/core/numeric.py", line 413, in count_nonzero
    return a.astype(np.bool_, copy=False).sum(axis=axis, dtype=np.intp)
ValueError: invalid literal for int() with base 10: ''
```

--- a/numpy/core/numeric.py
+++ b/numpy/core/numeric.py
@@ -411,33 +411,13 @@ def count_nonzero(a, axis=None):
 
     a = asanyarray(a)
 
-    if a.dtype == bool:
-        return a.sum(axis=axis, dtype=np.intp)
-
-    if issubdtype(a.dtype, np.number):
-        return (a != 0).sum(axis=axis, dtype=np.intp)
-
-    if issubdtype(a.dtype, np.character):
-        nullstr = a.dtype.type('')
-        return (a != nullstr).sum(axis=axis, dtype=np.intp)
-
-    axis = asarray(normalize_axis_tuple(axis, a.ndim))
-    counts = np.apply_along_axis(multiarray.count_nonzero, axis[0], a)
-
-    if axis.size == 1:
-        return counts.astype(np.intp, copy=False)
+    # TODO: this works around .astype(bool) not working properly (gh-9847)
+    if np.issubdtype(a.dtype, np.character):
+        a_bool = a != a.dtype.type()
     else:
-        # for subsequent axis numbers, that number decreases
-        # by one in this new 'counts' array if it was larger
-        # than the first axis upon which 'count_nonzero' was
-        # applied but remains unchanged if that number was
-        # smaller than that first axis
-        #
-        # this trick enables us to perform counts on object-like
-        # elements across multiple axes very quickly because integer
-        # addition is very well optimized
-        return counts.sum(axis=tuple(axis[1:] - (
-            axis[1:] > axis[0])), dtype=np.intp)
+        a_bool = a.astype(np.bool_, copy=False)
+
+    return a_bool.sum(axis=axis, dtype=np.intp)
 
 
 def asarray(a, dtype=None, order=None):

